File: ros_ws/src/cpp_pubsub/src/subscriber_member_function.cpp
#include <memory>
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>

using std::placeholders::_1;

class MinimalSubscriber : public rclcpp::Node
{
	public:
		MinimalSubscriber()
			: Node("minimal_subscriber")
		{
			constexpr size_t max_queue_length { 10 };
			m_subscription = this->create_subscription<std_msgs::msg::String>(
				"topic", max_queue_length, std::bind(&MinimalSubscriber::topic_callback, this, _1)
			);
		}

	private:
		void topic_callback(const std_msgs::msg::String& message) const
		{
			RCLCPP_INFO(this->get_logger(), "I heard: '%s'", message.data.c_str());
		}

		rclcpp::Subscription<std_msgs::msg::String>::SharedPtr m_subscription;
};

int main(int argc, char* argv[])
{
	rclcpp::init(argc, argv);
	rclcpp::spin(std::make_shared<MinimalSubscriber>());
	rclcpp::shutdown();
	return 0;
}

File: ros_ws/src/py_pubsub/py_pubsub/subscriber_member_function.py
import rclpy
from rclpy.node import Node

from std_msgs.msg import String


class MinimalSubscriber(Node):
    def __init__(self):
        super(MinimalSubscriber, self).__init__('minimal_subscriber')
        max_queue_length = 10
        self.subscription = self.create_subscription(
            String, 'topic', self.listener_callback, max_queue_length
        )

    def listener_callback(self, message: String):
        self.get_logger().info(f'I heard: "{message.data}"')


def main(args=None):
    rclpy.init(args=args)

    minimal_subscriber = MinimalSubscriber()

    rclpy.spin(minimal_subscriber)

    minimal_subscriber.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
